// @angular/fire/esm5/analytics/analytics.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __assign, __read } from "tslib";
import { Injectable, Optional, NgZone, ComponentFactoryResolver, Inject, PLATFORM_ID, Injector, NgModuleFactory } from '@angular/core';
import { from, Observable, of } from 'rxjs';
import { filter, withLatestFrom, switchMap, map, tap, pairwise, startWith, groupBy, mergeMap, observeOn } from 'rxjs/operators';
import { Router, NavigationEnd, ActivationEnd, ROUTES } from '@angular/router';
import { ɵAngularFireSchedulers } from '@angular/fire';
import { AngularFireAnalytics, DEBUG_MODE } from './analytics';
import { Title } from '@angular/platform-browser';
import { isPlatformBrowser, isPlatformServer } from '@angular/common';
import * as i0 from "@angular/core";
import * as i1 from "./analytics";
import * as i2 from "@angular/router";
import * as i3 from "@angular/platform-browser";
/** @type {?} */
var FIREBASE_EVENT_ORIGIN_KEY = 'firebase_event_origin';
/** @type {?} */
var FIREBASE_PREVIOUS_SCREEN_CLASS_KEY = 'firebase_previous_class';
/** @type {?} */
var FIREBASE_PREVIOUS_SCREEN_INSTANCE_ID_KEY = 'firebase_previous_id';
/** @type {?} */
var FIREBASE_PREVIOUS_SCREEN_NAME_KEY = 'firebase_previous_screen';
/** @type {?} */
var FIREBASE_SCREEN_CLASS_KEY = 'firebase_screen_class';
/** @type {?} */
var FIREBASE_SCREEN_INSTANCE_ID_KEY = 'firebase_screen_id';
/** @type {?} */
var FIREBASE_SCREEN_NAME_KEY = 'firebase_screen';
/** @type {?} */
var OUTLET_KEY = 'outlet';
/** @type {?} */
var PAGE_PATH_KEY = 'page_path';
/** @type {?} */
var PAGE_TITLE_KEY = 'page_title';
/** @type {?} */
var SCREEN_CLASS_KEY = 'screen_class';
/** @type {?} */
var SCREEN_NAME_KEY = 'screen_name';
/** @type {?} */
var SCREEN_VIEW_EVENT = 'screen_view';
/** @type {?} */
var EVENT_ORIGIN_AUTO = 'auto';
/** @type {?} */
var DEFAULT_SCREEN_CLASS = '???';
/** @type {?} */
var NG_PRIMARY_OUTLET = 'primary';
/** @type {?} */
var SCREEN_INSTANCE_DELIMITER = '#';
/** @type {?} */
var ANNOTATIONS = '__annotations__';
var ScreenTrackingService = /** @class */ (function () {
    function ScreenTrackingService(analytics, router, title, componentFactoryResolver, platformId, debugModeEnabled, zone, injector) {
        var _this = this;
        if (!router || !isPlatformBrowser(platformId)) {
            return this;
        }
        zone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var activationEndEvents = router.events.pipe(filter((/**
             * @param {?} e
             * @return {?}
             */
            function (e) { return e instanceof ActivationEnd; })));
            /** @type {?} */
            var navigationEndEvents = router.events.pipe(filter((/**
             * @param {?} e
             * @return {?}
             */
            function (e) { return e instanceof NavigationEnd; })));
            _this.disposable = navigationEndEvents.pipe(withLatestFrom(activationEndEvents), switchMap((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b, _c, _d, _e, _f;
                var _g = __read(_a, 2), navigationEnd = _g[0], activationEnd = _g[1];
                // SEMVER: start using optional chains and nullish coalescing once we support newer typescript
                /** @type {?} */
                var page_path = navigationEnd.url;
                /** @type {?} */
                var screen_name = activationEnd.snapshot.routeConfig && activationEnd.snapshot.routeConfig.path || page_path;
                /** @type {?} */
                var params = (_b = {},
                    _b[SCREEN_NAME_KEY] = screen_name,
                    _b[PAGE_PATH_KEY] = page_path,
                    _b[FIREBASE_EVENT_ORIGIN_KEY] = EVENT_ORIGIN_AUTO,
                    _b[FIREBASE_SCREEN_NAME_KEY] = screen_name,
                    _b[OUTLET_KEY] = activationEnd.snapshot.outlet,
                    _b);
                if (title) {
                    params[PAGE_TITLE_KEY] = title.getTitle();
                }
                /** @type {?} */
                var component = activationEnd.snapshot.component;
                /** @type {?} */
                var routeConfig = activationEnd.snapshot.routeConfig;
                /** @type {?} */
                var loadChildren = routeConfig && routeConfig.loadChildren;
                // TODO figure out how to handle minification
                if (typeof loadChildren === "string") {
                    // SEMVER: this is the older lazy load style "./path#ClassName", drop this when we drop old ng
                    // TODO is it worth seeing if I can look up the component factory selector from the module name?
                    // it's lazy so it's not registered with componentFactoryResolver yet... seems a pain for a depreciated style
                    return of(__assign(__assign({}, params), (_c = {}, _c[SCREEN_CLASS_KEY] = loadChildren.split('#')[1], _c)));
                }
                else if (typeof component === 'string') {
                    return of(__assign(__assign({}, params), (_d = {}, _d[SCREEN_CLASS_KEY] = component, _d)));
                }
                else if (component) {
                    /** @type {?} */
                    var componentFactory = componentFactoryResolver.resolveComponentFactory(component);
                    return of(__assign(__assign({}, params), (_e = {}, _e[SCREEN_CLASS_KEY] = componentFactory.selector, _e)));
                }
                else if (loadChildren) {
                    /** @type {?} */
                    var loadedChildren = loadChildren();
                    /** @type {?} */
                    var loadedChildren$ = (loadedChildren instanceof Observable) ? loadedChildren : from(Promise.resolve(loadedChildren));
                    return loadedChildren$.pipe(map((/**
                     * @param {?} lazyModule
                     * @return {?}
                     */
                    function (lazyModule) {
                        var _a, _b, _c;
                        if (lazyModule instanceof NgModuleFactory) {
                            // AOT create an injector
                            /** @type {?} */
                            var moduleRef = lazyModule.create(injector);
                            // INVESTIGATE is this the right way to get at the matching route?
                            /** @type {?} */
                            var routes = moduleRef.injector.get(ROUTES);
                            /** @type {?} */
                            var component_1 = routes[0][0].component;
                            try {
                                /** @type {?} */
                                var componentFactory = moduleRef.componentFactoryResolver.resolveComponentFactory((/** @type {?} */ (component_1)));
                                return __assign(__assign({}, params), (_a = {}, _a[SCREEN_CLASS_KEY] = componentFactory.selector, _a));
                            }
                            catch (_) {
                                return __assign(__assign({}, params), (_b = {}, _b[SCREEN_CLASS_KEY] = DEFAULT_SCREEN_CLASS, _b));
                            }
                        }
                        else {
                            // JIT look at the annotations
                            // INVESTIGATE are there public APIs for this stuff?
                            /** @type {?} */
                            var declarations = [].concat.apply([], (lazyModule[ANNOTATIONS] || []).map((/**
                             * @param {?} f
                             * @return {?}
                             */
                            function (f) { return f.declarations; })));
                            /** @type {?} */
                            var selectors = [].concat.apply([], declarations.map((/**
                             * @param {?} c
                             * @return {?}
                             */
                            function (c) { return (c[ANNOTATIONS] || []).map((/**
                             * @param {?} f
                             * @return {?}
                             */
                            function (f) { return f.selector; })); })));
                            // should I just be grabbing the selector like this or should i match against the route component?
                            //   const routerModule = lazyModule.ngInjectorDef.imports.find(i => i.ngModule && ....);
                            //   const route = routerModule.providers[0].find(p => p.provide == ROUTES).useValue[0];
                            return __assign(__assign({}, params), (_c = {}, _c[SCREEN_CLASS_KEY] = selectors[0] || DEFAULT_SCREEN_CLASS, _c));
                        }
                    })));
                }
                else {
                    return of(__assign(__assign({}, params), (_f = {}, _f[SCREEN_CLASS_KEY] = DEFAULT_SCREEN_CLASS, _f)));
                }
            })), map((/**
             * @param {?} params
             * @return {?}
             */
            function (params) {
                var _a;
                return (__assign((_a = {}, _a[FIREBASE_SCREEN_CLASS_KEY] = params[SCREEN_CLASS_KEY], _a[FIREBASE_SCREEN_INSTANCE_ID_KEY] = getScreenInstanceID(params), _a), params));
            })), tap((/**
             * @param {?} params
             * @return {?}
             */
            function (params) {
                var _a, _b;
                // TODO perhaps I can be smarter about this, bubble events up to the nearest outlet?
                if (params[OUTLET_KEY] == NG_PRIMARY_OUTLET) {
                    analytics.setCurrentScreen(params[SCREEN_NAME_KEY]);
                    analytics.updateConfig((_a = {},
                        _a[PAGE_PATH_KEY] = params[PAGE_PATH_KEY],
                        _a[SCREEN_CLASS_KEY] = params[SCREEN_CLASS_KEY],
                        _a));
                    if (title) {
                        analytics.updateConfig((_b = {}, _b[PAGE_TITLE_KEY] = params[PAGE_TITLE_KEY], _b));
                    }
                }
            })), groupBy((/**
             * @param {?} params
             * @return {?}
             */
            function (params) { return params[OUTLET_KEY]; })), mergeMap((/**
             * @param {?} group
             * @return {?}
             */
            function (group) { return group.pipe(startWith(undefined), pairwise()); })), map((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b;
                var _c = __read(_a, 2), prior = _c[0], current = _c[1];
                return prior ? __assign((_b = {}, _b[FIREBASE_PREVIOUS_SCREEN_CLASS_KEY] = prior[SCREEN_CLASS_KEY], _b[FIREBASE_PREVIOUS_SCREEN_NAME_KEY] = prior[SCREEN_NAME_KEY], _b[FIREBASE_PREVIOUS_SCREEN_INSTANCE_ID_KEY] = prior[FIREBASE_SCREEN_INSTANCE_ID_KEY], _b), (/** @type {?} */ (current))) : (/** @type {?} */ (current));
            })), tap((/**
             * @param {?} params
             * @return {?}
             */
            function (params) { return debugModeEnabled && console.info(SCREEN_VIEW_EVENT, params); })), tap((/**
             * @param {?} params
             * @return {?}
             */
            function (params) { return zone.runOutsideAngular((/**
             * @return {?}
             */
            function () { return analytics.logEvent(SCREEN_VIEW_EVENT, params); })); }))).subscribe();
        }));
    }
    /**
     * @return {?}
     */
    ScreenTrackingService.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.disposable) {
            this.disposable.unsubscribe();
        }
    };
    ScreenTrackingService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'any'
                },] }
    ];
    /** @nocollapse */
    ScreenTrackingService.ctorParameters = function () { return [
        { type: AngularFireAnalytics },
        { type: Router, decorators: [{ type: Optional }] },
        { type: Title, decorators: [{ type: Optional }] },
        { type: ComponentFactoryResolver },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [DEBUG_MODE,] }] },
        { type: NgZone },
        { type: Injector }
    ]; };
    /** @nocollapse */ ScreenTrackingService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ScreenTrackingService_Factory() { return new ScreenTrackingService(i0.ɵɵinject(i1.AngularFireAnalytics), i0.ɵɵinject(i2.Router, 8), i0.ɵɵinject(i3.Title, 8), i0.ɵɵinject(i0.ComponentFactoryResolver), i0.ɵɵinject(i0.PLATFORM_ID), i0.ɵɵinject(i1.DEBUG_MODE, 8), i0.ɵɵinject(i0.NgZone), i0.ɵɵinject(i0.INJECTOR)); }, token: ScreenTrackingService, providedIn: "any" });
    return ScreenTrackingService;
}());
export { ScreenTrackingService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ScreenTrackingService.prototype.disposable;
}
var UserTrackingService = /** @class */ (function () {
    // TODO a user properties injector
    function UserTrackingService(analytics, zone, platformId) {
        var _this = this;
        /** @type {?} */
        var schedulers = new ɵAngularFireSchedulers(zone);
        if (!isPlatformServer(platformId)) {
            zone.runOutsideAngular((/**
             * @return {?}
             */
            function () {
                // @ts-ignore zap the import in the UMD
                _this.disposable = from(import('firebase/auth')).pipe(observeOn(schedulers.outsideAngular), switchMap((/**
                 * @return {?}
                 */
                function () { return analytics.app; })), map((/**
                 * @param {?} app
                 * @return {?}
                 */
                function (app) { return app.auth(); })), switchMap((/**
                 * @param {?} auth
                 * @return {?}
                 */
                function (auth) { return new Observable(auth.onAuthStateChanged.bind(auth)); })), switchMap((/**
                 * @param {?} user
                 * @return {?}
                 */
                function (user) { return analytics.setUserId(user ? user.uid : (/** @type {?} */ (null))); }))).subscribe();
            }));
        }
    }
    /**
     * @return {?}
     */
    UserTrackingService.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.disposable) {
            this.disposable.unsubscribe();
        }
    };
    UserTrackingService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'any'
                },] }
    ];
    /** @nocollapse */
    UserTrackingService.ctorParameters = function () { return [
        { type: AngularFireAnalytics },
        { type: NgZone },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    /** @nocollapse */ UserTrackingService.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserTrackingService_Factory() { return new UserTrackingService(i0.ɵɵinject(i1.AngularFireAnalytics), i0.ɵɵinject(i0.NgZone), i0.ɵɵinject(i0.PLATFORM_ID)); }, token: UserTrackingService, providedIn: "any" });
    return UserTrackingService;
}());
export { UserTrackingService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UserTrackingService.prototype.disposable;
}
// this is an INT64 in iOS/Android but use INT32 cause javascript
/** @type {?} */
var nextScreenInstanceID = Math.floor(Math.random() * (Math.pow(2, 32) - 1)) - Math.pow(2, 31);
/** @type {?} */
var knownScreenInstanceIDs = {};
/** @type {?} */
var getScreenInstanceID = (/**
 * @param {?} params
 * @return {?}
 */
function (params) {
    // unique the screen class against the outlet name
    /** @type {?} */
    var screenInstanceKey = [
        params[SCREEN_CLASS_KEY],
        params[OUTLET_KEY]
    ].join(SCREEN_INSTANCE_DELIMITER);
    if (knownScreenInstanceIDs.hasOwnProperty(screenInstanceKey)) {
        return knownScreenInstanceIDs[screenInstanceKey];
    }
    else {
        /** @type {?} */
        var ret = nextScreenInstanceID++;
        knownScreenInstanceIDs[screenInstanceKey] = ret;
        return ret;
    }
});
var ɵ0 = getScreenInstanceID;
export { ɵ0 };
