// @angular/fire/esm5/database/object/create-reference.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { map } from 'rxjs/operators';
import { createObjectSnapshotChanges } from './snapshot-changes';
/**
 * @template T
 * @param {?} query
 * @param {?} afDatabase
 * @return {?}
 */
export function createObjectReference(query, afDatabase) {
    return {
        query: query,
        snapshotChanges: /**
         * @template T
         * @return {?}
         */
        function () {
            return createObjectSnapshotChanges(query, afDatabase.schedulers.outsideAngular)().pipe((/** @type {?} */ (afDatabase.keepUnstableUntilFirst)));
        },
        update: /**
         * @param {?} data
         * @return {?}
         */
        function (data) { return (/** @type {?} */ (query.ref.update((/** @type {?} */ (data))))); },
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) { return (/** @type {?} */ (query.ref.set(data))); },
        remove: /**
         * @return {?}
         */
        function () { return (/** @type {?} */ (query.ref.remove())); },
        valueChanges: /**
         * @template T
         * @return {?}
         */
        function () {
            /** @type {?} */
            var snapshotChanges$ = createObjectSnapshotChanges(query, afDatabase.schedulers.outsideAngular)();
            return snapshotChanges$.pipe(afDatabase.keepUnstableUntilFirst, map((/**
             * @param {?} action
             * @return {?}
             */
            function (action) { return action.payload.exists() ? (/** @type {?} */ (action.payload.val())) : null; })));
        },
    };
}
